### DISCUSSION TITLE
Is Supabase Function (Postgres Function) safe for sql injection?

### DISCUSSION THREAD

**Discussion by hereAlexT** [supabase/supabase#25917] [Questions] [answered]
I need to use `transaction` for some operations; the only method I found to use `transaction` in Supabase is the Supabase Function, which is also the Postgres Function. 

In the Supabase Function, I have to write SQL or PL/PG SQL. 

Is it safe to use the Supabase function directly without worrying about the injection? 
If its not secure, what can i do to prevent injection and other attacks?

**Comment by GaryAustin1** [comment]
It is safe as long as you don't use the parameters to generate formatted SQL code and use something like EXECUTE to run that code.

**Reply by hereAlexT** [reply]
Thanks for your reply!
How do I make it safe to pass parameters to the function? 
Is there any other way to run multiple commands to obtain the same effect as a transaction that applies the ACID principle?

**Reply by GaryAustin1** [reply, marked as answer]
To do normal SQL in functions you don't need to do anything with the data.  It can be used for where clauses and as inserted data with no injection issues.  

Functions are the way to do transactions.

**Reply by GaryAustin1** [reply]
For instance this is safe from injection:
```SQL
create or replace function add_planet(name text)
returns bigint
language plpgsql
as $$
declare
  new_row bigint;
begin
  insert into planets(name)
  values (add_planet.name)
  returning id into new_row;

  return new_row;
end;
$$;
```

**Reply by hereAlexT** [reply]
Thank you. Appreciate it!
